drivers/rtc/rtc_mchp_g1.c

```diff
@@ -0,0 +1,964 @@
+{
+	const struct rtc_mchp_dev_config *const cfg = dev->config;
+
+	/* Calculate the calibration value in parts per billion */
+	int32_t correction = calibration / (RTC_CALIB_PARTS_PER_BILLION / cfg->cal_constant);
+	uint32_t abs_correction = abs(correction);
+
+	LOG_DBG("Correction: %d, Absolute: %d, Calibration: %d", correction, abs_correction,
+		calibration);
+
+	/* Check if correction value is 0 */
+	if (abs_correction == 0) {
+		rtc_set_calibration_value(cfg->regs, 0, 0);
+	}
+
+	/* Check if correction value out of range */
+	if (abs_correction > RTC_CALIBRATE_PPB_MAX) {
+		LOG_ERR("The RTC calibration %d result in an out of range value %d", calibration,
+			abs_correction);
+		return -EINVAL;
+	}
+
+	/* Set the correction value into RTC calib register */
+	rtc_set_calibration_value(cfg->regs, abs_correction, correction);
+
+	return 0;
+}
+
+static int rtc_mchp_get_calibration(const struct device *dev, int32_t *calibration)
+{
+	const struct rtc_mchp_dev_config *const cfg = dev->config;
+	int32_t correction;
+	uint8_t correction_sign = 0;
+	int status;
+
+	/* Check if the calibration pointer is NULL */
+	if (calibration == NULL) {
+		return -EINVAL;
+	}
+
+	/* Retrieve the correction value from the hardware register */
+	status = rtc_get_calibration_value(cfg->regs, &correction, &correction_sign);
+	if (status < 0) {
+		return -EINVAL;
+	}
+
+	/* Calculate the calibration value based on the correction value */
+	if (correction == 0) {
+		*calibration = 0;
+	} else {
+		*calibration =
+			((int64_t)correction * RTC_CALIB_PARTS_PER_BILLION) / cfg->cal_constant;
+	}
+
+	/* Adjust the calibration value based on the sign bit */
+	if (correction_sign == 1) {
+		*calibration *= -1;
+	}
+
+	return 0;
+}
+#endif /* CONFIG_RTC_CALIBRATION */
+
+/*
+ * Initialize the RTC (Real-Time Clock) for the Microchip device.
+ *
+ * This function initializes the RTC hardware by enabling the clock, performing a software
+ * reset, setting the prescaler, enabling the clock calendar mode, and enabling the RTC. If
+ * RTC alarm configuration is enabled, it also configures the IRQ for the RTC peripheral.
+ */
+static int rtc_mchp_init(const struct device *dev)
+{
+	struct rtc_mchp_dev_data *data = dev->data;
+	const struct rtc_mchp_dev_config *const cfg = dev->config;
+	int ret = 0;
+
+	/* On Oscillator clock for RTC */
+	ret = clock_control_on(cfg->rtc_clock.clock_dev, cfg->rtc_clock.rtcclk_sys);
+	if ((ret != 0) && (ret != -EALREADY)) {
+		LOG_ERR("Failed to enable the osc32k clock for RTC: %d", ret);
+		return ret;
+	}
+
+	/* On Main clock for RTC */
+	ret = clock_control_on(cfg->rtc_clock.clock_dev, cfg->rtc_clock.mclk_sys);
+	if ((ret != 0) && (ret != -EALREADY)) {
+		LOG_ERR("Failed to enable the MCLK for RTC: %d", ret);
+		return ret;
+	}
+
+	/* Initialize mutex for RTC data structure */
+	RTC_DATA_LOCK_INIT(&data->lock);
+
+	/* Perform a software reset on the RTC peripheral */
+	rtc_swrst(cfg->regs);
```

Comment:
Will this action reset the time of the rtc? RTCs are supposed to survive a reboot, and any other event which does not physically take away power if possible.